haystack/nodes/preprocessor/preprocessor.py

```diff
@@ -315,116 +567,121 @@ def split(
-            elif split_by == "sentence":
-                if self.add_page_number:
-                    # SentenceTokenizer will remove "\f" if it is at the end of a sentence, so substituting it in these
-                    # cases for "[NEW_PAGE]" to don't lose any page breaks.
-                    text = self._substitute_page_breaks(text)
-                elements = self._split_sentences(text)
-            elif split_by == "word":
-                elements = text.split(" ")
-            else:
-                raise NotImplementedError(
-                    "PreProcessor only supports 'passage', 'sentence' or 'word' split_by options."
-                )
-
-            # concatenate individual elements based on split_length & split_stride
-            if split_overlap:
-                segments = windowed(elements, n=split_length, step=split_length - split_overlap)
-            else:
-                segments = windowed(elements, n=split_length, step=split_length)
             text_splits = []
             splits_pages = []
             cur_page = 1
-            for seg in segments:
-                current_units = [unit for unit in seg if unit is not None]
-                txt = " ".join(current_units)
-                if len(txt) > 0:
-                    text_splits.append(txt)
-                    splits_pages.append(cur_page)
+            # Loop through each paragraph/passage. If pre_split_paragraphs is set to False, then this for loop will only run once
+            for para in paras:
+                if split_by == "passage":
+                    elements = self._split_paragraphs(para, merge_short, merge_lowercase)
+                elif split_by == "sentence":
                     if self.add_page_number:
-                        processed_units = current_units[: split_length - split_overlap]
-                        num_page_breaks = sum(processed_unit.count("\f") for processed_unit in processed_units)
-                        cur_page += num_page_breaks
+                        # SentenceTokenizer will remove "\f" if it is at the end of a sentence, so substituting it in these
+                        # cases for "[NEW_PAGE]" to don't lose any page breaks.
+                        para = self._substitute_page_breaks(para)
+                    elements = self._split_sentences(para)
+
+                    for i, sentence in enumerate(elements):
+                        if len(paras) == 1:
+                            elements[i] = sentence.replace("\n\n", " ")
+                        if self.add_page_number and sentence.startswith("[NEW_PAGE]"):
+                            elements[i] = sentence.replace("[NEW_PAGE]", "\f")
+                elif split_by == "word":
+                    elements = para.replace("\n\n", " ").strip().split(" ")
+                else:
+                    raise NotImplementedError(
+                        "PreProcessor only supports 'passage', 'sentence' or 'word' split_by options."
+                    )
```

Comment:
I'd say let's keep it for safety. Shouldn't ever be triggered, but if we change the code again in the future it might. I'd change the error into a ValueError though.